### DISCUSSION TITLE
Movement prediction

### DISCUSSION THREAD

**Discussion by Gamer1120** [runelite/runelite#15306] [Ideas]
During the RuneFest RS3 keynote, it was shown how they wanted to implement movement prediction on the client side. When you click somewhere, you will go there instantly client-side, instead of waiting for a server response. This would be pretty neat to have.

**Comment by notjuanortiz** [comment]
That's a pretty common feature in MMOs used to mitigate high latency issues. I'm not sure if this is out of the scope of this project BUT if we were to implement it, this is how I would do it.

1. Click on a random tile OR point on the map
2. Client sends an **asynchronous** remote procedural call (RPC) message to the server
2. Client moves the player towards the selected tile/point
3. Client receives RPC response from server containing the actual position of the player
4. Calculate points between the server-dictated position (SDP) and the client's current position using linear spline interpolation. 
5. Render the player at one of the interpolation points. This will obviously require some playtesting to determine what looks the best.

If we're going for instantaneous correction it'll lead to player snapping, which is when the player abruptly moves to the SDP. Since we're calculating multiple interpolated points we could create a smooth transition between them. Although this would work fine, there are certain instances where player snapping makes sense. For instance, when the client's position is outside of the SDP's render distance (when teleporting or moving up or down a ladder).

**Comment by notjuanortiz** [comment]
After some investigation it looks like the parts of the client that are crucial to the implementation of this feature are inaccessible. Whichever system is responsible for requesting movement updates from the server is inaccessible. Similarly, whatever class is responsible for rendering the player model controlled by the client itself, is also inaccessible.

I was pointed towards the GPU plugin to begin to understand how rendering manipulation was done, only to discover the plugin is actually rendering a new canvas on top of the preexisting one and **not** manipulating the preexisting one.

As of now, due to restricted access to portions of client, _my initial approach will not work_.
